FlowCrypt/Controllers/Msg/MsgViewController.swift

```diff
@@ -7,179 +7,223 @@ import MBProgressHUD
-        btnBack.frame = Constants.uiBarButtonItemFrame
-        btnBack.addTarget(self, action: #selector(btnBackTap), for: .touchUpInside)
-        
-        self.navigationItem.leftBarButtonItem = UIBarButtonItem(customView: btnBack)
     }
-    
-    func renderMsgBody(_ text: String, color: UIColor = UIColor.black) {
-        self.lblBody.text = text
-        self.lblBody.textColor = color
-        self.lblBody.sizeToFit()
-        self.scrollView.contentSize = self.scrollViewContent.frame.size;
+
+    private func setupNavigationBar() {
+        let infoInput = NavigationBarItemsView.Input(
+            image: UIImage(named: "help_icn"),
+            action: (self, #selector(handleInfoTap))
+        )
+
+        let archiveInput = NavigationBarItemsView.Input(
+            image: UIImage(named: "archive"),
+            action: (self, #selector(handleArchiveTap))
+        )
+
+        let trashInput = NavigationBarItemsView.Input(
+            image: UIImage(named: "trash"),
+            action: (self, #selector(handleTrashTap))
+        )
+
+        let mailInput = NavigationBarItemsView.Input(
+            image: UIImage(named: "mail"),
+            action: (self, #selector(handleMailTap))
+        )
+
+        let buttons = input?.path == MailDestination.Gmail.trash.path
+            ? [infoInput, archiveInput, mailInput]
+            : [infoInput, archiveInput, trashInput, mailInput]
+
+        navigationItem.rightBarButtonItem = NavigationBarItemsView(with: buttons)
+        navigationItem.leftBarButtonItem = NavigationBarActionButton(UIImage(named: "arrow-left-c")) { [weak self] in
+            self?.navigationController?.popViewController(animated: true)
+        }
     }
-    
-    func markAsReadIfNotAlreadyMarked() {
-        if !objMessage.flags.isSuperset(of: MCOMessageFlag.seen) {
-            self.objMessage.flags.formUnion(MCOMessageFlag.seen)
-            let _ = Imap.instance.markAsRead(message: self.objMessage, folder: self.path) // async, do not await
-            guard let delegate = self.delegate else { return }
-            delegate.movedOrUpdated(objMessage: self.objMessage)            
+
+    private func renderMsgBody(_ text: String, color: UIColor = UIColor.black) {
+        lblBody.text = text
+        lblBody.textColor = color
+        lblBody.sizeToFit()
+        scrollView.contentSize = scrollViewContent.frame.size
+    }
+}
+
+// MARK: - Message
+extension MsgViewController {
+    private func fetchMessage() {
+        guard let input = input else { return }
+        showSpinner(Language.loading, isUserInteractionEnabled: true)
+        imap.fetchMsg(message: input.objMessage, folder: input.path)
+            .then { [weak self] data -> Promise<CoreRes.ParseDecryptMsg> in
+                guard let self = self else { return Promise(ImapError.general) }
+                self.input?.bodyMessage = data
+                return self.db.save(message: data, isEmail: true)
```

Comment:
This looks wrong:
 - there is no need to save the message anywhere (yet)
 - the `db.save` method is not saving anything anyway, it is only loading private keys from db and decrypting
 - it should be VERY obvious that we are decrypting a message, `self.db.save` is a complete mismatch in terms of naming

Just decrypt it right here, as before.